carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/utils/FeatureFileGeneratorUtils.java

```diff
@@ -0,0 +1,452 @@
+        }
+        Document document;
+        if (manifest != null && manifest.exists()) {
+            try (InputStream manifestFileStream = new FileInputStream(manifest)) {
+                document = documentBuilder.parse(manifestFileStream);
+            } catch (SAXException e) {
+                throw new SAXException("Unable to parse feature manifest", e);
+            } catch (IOException e) {
+                throw new IOException("Unable to load feature manifest", e);
+            }
+        } else {
+            document = documentBuilder.newDocument();
+        }
+        return document;
+    }
+
+    /**
+     * Cross check plugins given in the manifest file against the plugins configured in the pom.xml file. Returns a
+     * list of bundles found in the pom.xml but not in the manifest file.
+     *
+     * @param bundles  list of bundles configured in the pom.xml
+     * @param document Document representing the give manifest
+     * @return ArrayList&lt;Bundle&gt; missing plugins
+     */
+    private static ArrayList<Bundle> getMissingPlugins(List<Bundle> bundles, Document document) {
+        if (bundles == null || bundles.size() == 0) {
+            return new ArrayList<>();
+        }
+        HashMap<String, Bundle> missingPlugins = new HashMap<>();
+        for (Bundle bundle : bundles) {
+            missingPlugins.put(bundle.getArtifactId(), bundle);
+        }
+
+        NodeList existingPlugins = document.getDocumentElement().getElementsByTagName("plugin");
+
+        if (existingPlugins != null) {
+            for (int i = 0; i < existingPlugins.getLength(); i++) {
+                Node node = existingPlugins.item(i);
+                Node namedItem = node.getAttributes().getNamedItem("id");
+                if (namedItem != null && namedItem.getTextContent() != null &&
+                        missingPlugins.containsKey(namedItem.getTextContent())) {
+                    missingPlugins.remove(namedItem.getTextContent());
+                }
+            }
+        }
+
+        return new ArrayList<>(missingPlugins.values());
+    }
+
+    /**
+     * Cross check import bundles/import features in the given manifest file against the plugins configured in the
+     * pom.xml file. Returns a list of import bundles/import features found in the pom.xml but not in the manifest file.
+     *
+     * @param processedImportItemsList list of import plugins/import features configured in the pom.xml
+     * @param document                 Document representing the give manifest
+     * @param itemType                 String type, either "feature" or "plugin"
+     * @return ArrayList<Feature>
+     */
+    private static ArrayList<Feature> getMissingImportFeatures(List<Feature> processedImportItemsList, Document document,
+                                                               String itemType) {
+        if (processedImportItemsList == null) {
+            return new ArrayList<>();
```

Comment:
What about making the return statement use "Collections.emptyList()" instead of instantiating a new ArrayList? I hope the returned list is not modified anywhere else in the code.